File: scripts/extract_frames.m
#import <AVFoundation/AVFoundation.h>
#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc < 3) {
            fprintf(stderr, "usage: extract_frames VIDEO OUTPUT_DIR [interval_seconds]\n");
            return 2;
        }

        NSString *inputPath = [NSString stringWithUTF8String:argv[1]];
        NSString *outputPath = [NSString stringWithUTF8String:argv[2]];
        double interval = argc >= 4 ? atof(argv[3]) : 1.0;
        [[NSFileManager defaultManager] createDirectoryAtPath:outputPath
                                  withIntermediateDirectories:YES
                                                   attributes:nil
                                                        error:nil];

        AVURLAsset *asset = [AVURLAsset URLAssetWithURL:[NSURL fileURLWithPath:inputPath] options:nil];
        double duration = CMTimeGetSeconds(asset.duration);
        AVAssetTrack *track = [[asset tracksWithMediaType:AVMediaTypeVideo] firstObject];
        if (track == nil) {
            fprintf(stderr, "no video track\n");
            return 3;
        }
        CGSize size = track.naturalSize;
        printf("duration=%.3f width=%d height=%d interval=%.3f\n",
               duration, (int)size.width, (int)size.height, interval);

        AVAssetImageGenerator *generator = [[AVAssetImageGenerator alloc] initWithAsset:asset];
        generator.appliesPreferredTrackTransform = YES;
        generator.requestedTimeToleranceBefore = kCMTimeZero;
        generator.requestedTimeToleranceAfter = kCMTimeZero;

        int index = 0;
        for (double timestamp = 0.0; timestamp <= duration; timestamp += interval, index++) {
            @autoreleasepool {
                NSError *error = nil;
                CGImageRef image = [generator copyCGImageAtTime:CMTimeMakeWithSeconds(timestamp, 600)
                                                     actualTime:nil
                                                          error:&error];
                if (image == nil) {
                    fprintf(stderr, "frame %d at %.3fs failed: %s\n",
                            index, timestamp, error.localizedDescription.UTF8String);
                    continue;
                }

                NSBitmapImageRep *bitmap = [[NSBitmapImageRep alloc] initWithCGImage:image];
                NSData *png = [bitmap representationUsingType:NSBitmapImageFileTypePNG properties:@{}];
                NSString *name = [NSString stringWithFormat:@"frame_%05d_%09.3f.png", index, timestamp];
                [png writeToFile:[outputPath stringByAppendingPathComponent:name] atomically:YES];
                CGImageRelease(image);
            }
        }

        printf("frames=%d\n", index);
    }
    return 0;
}

File: scripts/extract_frames.swift
import AVFoundation
import AppKit
import Foundation

guard CommandLine.arguments.count >= 3 else {
    fputs("usage: extract_frames.swift VIDEO OUTPUT_DIR [interval_seconds]\n", stderr)
    exit(2)
}

let inputURL = URL(fileURLWithPath: CommandLine.arguments[1])
let outputURL = URL(fileURLWithPath: CommandLine.arguments[2], isDirectory: true)
let interval = CommandLine.arguments.count >= 4 ? (Double(CommandLine.arguments[3]) ?? 1.0) : 1.0

try FileManager.default.createDirectory(at: outputURL, withIntermediateDirectories: true)

let asset = AVURLAsset(url: inputURL)
let duration = try await asset.load(.duration)
let seconds = CMTimeGetSeconds(duration)
let tracks = try await asset.loadTracks(withMediaType: .video)
guard let track = tracks.first else {
    fputs("no video track\n", stderr)
    exit(3)
}
let naturalSize = try await track.load(.naturalSize)
print("duration=\(String(format: "%.3f", seconds)) width=\(Int(naturalSize.width)) height=\(Int(naturalSize.height)) interval=\(interval)")

let generator = AVAssetImageGenerator(asset: asset)
generator.appliesPreferredTrackTransform = true
generator.requestedTimeToleranceBefore = .zero
generator.requestedTimeToleranceAfter = .zero

var index = 0
var timestamp = 0.0
while timestamp <= seconds {
    autoreleasepool {
        let requestedTime = CMTime(seconds: timestamp, preferredTimescale: 600)
        do {
            let cgImage = try generator.copyCGImage(at: requestedTime, actualTime: nil)
            let bitmap = NSBitmapImageRep(cgImage: cgImage)
            guard let png = bitmap.representation(using: .png, properties: [:]) else {
                throw NSError(domain: "extract_frames", code: 1)
            }
            let name = String(format: "frame_%05d_%09.3f.png", index, timestamp)
            try png.write(to: outputURL.appendingPathComponent(name))
        } catch {
            fputs("frame \(index) at \(timestamp)s failed: \(error)\n", stderr)
        }
    }
    index += 1
    timestamp += interval
}

print("frames=\(index)")
